api/client_test.go

```diff
@@ -47,46 +75,347 @@ func Test_DefaultClientConfig_WithEnvVars(t *testing.T) {
+			want:   "status=foo",
+		},
+		{
+			name:   "multiple fields",
+			fields: fields{Status: "foo", Run: "bar"},
+			want:   "run=bar&status=foo",
+		},
+		{
+			name:   "multiple fields with include events",
+			fields: fields{Status: "foo", Run: "bar", IncludeEvents: true},
+			want:   "include=events&run=bar&status=foo",
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			q := &QueryParam{
+				IncludeEvents: tt.fields.IncludeEvents,
+				Status:        tt.fields.Status,
+				Run:           tt.fields.Run,
+			}
+			assert.Equalf(t, tt.want, q.Encode(), "Encode()")
+		})
+	}
+}
+
+func Test_StatusClient_Overall_BadResponse(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := fmt.Fprintf(w, "foo")
+		assert.NoError(t, err)
+	}))
+
+	defer server.Close()
+
+	clientConfig := BaseClientConfig()
+	clientConfig.URL = server.URL
+	c, err := NewClient(clientConfig, nil)
+	assert.NoError(t, err)
+
+	o, err := c.Status().Overall()
+	assert.Error(t, err)
+	assert.EqualValues(t, OverallStatus{}, o)
+}
+
+func Test_StatusClient_Overall_Error_BadResponse(t *testing.T) {
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusInternalServerError)
+	}))
+
+	defer server.Close()
+
+	clientConfig := BaseClientConfig()
+	clientConfig.URL = server.URL
+	c, err := NewClient(clientConfig, nil)
+	assert.NoError(t, err)
+
+	o, err := c.Status().Overall()
+	assert.Error(t, err)
+	assert.EqualValues(t, OverallStatus{}, o)
+}
+
+func Test_StatusClient_Overall_Error_WithErrorMessage(t *testing.T) {
+	e := &ErrorResponse{Error: &ErrorObject{Message: "foo"}}
+	bytes, err := json.Marshal(e)
+	assert.NotEmpty(t, bytes)
+	assert.NoError(t, err)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, err := w.Write(bytes)
+		assert.NoError(t, err)
+	}))
+
+	defer server.Close()
+
+	clientConfig := BaseClientConfig()
+	clientConfig.URL = server.URL
+	c, err := NewClient(clientConfig, nil)
+	assert.NoError(t, err)
+
+	o, err := c.Status().Overall()
+	assert.Error(t, err)
+	assert.EqualValues(t, OverallStatus{}, o)
+}
+
+func Test_StatusClient_Overall_Error_NoErrorMessage(t *testing.T) {
+	e := &ErrorResponse{}
+	bytes, err := json.Marshal(e)
+	assert.NotEmpty(t, bytes)
+	assert.NoError(t, err)
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusInternalServerError)
+		_, err := w.Write(bytes)
+		assert.NoError(t, err)
+	}))
+
+	defer server.Close()
+
+	clientConfig := BaseClientConfig()
+	clientConfig.URL = server.URL
+	c, err := NewClient(clientConfig, nil)
+	assert.NoError(t, err)
+
+	o, err := c.Status().Overall()
+	assert.Error(t, err)
+	assert.EqualValues(t, OverallStatus{}, o)
+}
+
+func Test_StatusClient_Overall(t *testing.T) {
```

Comment:
All these test functions for Overall Status client are really thorough! Since there's other client methods, I'm curious to know your thoughts on those. Note - not trying to imply that they should be tested here.

Reply:
If you are talking about Task() and Update(), then yes, I do plan on testing them as well, but I wanted to check with you because I think there's an opportunity to refactor them and change their signature a bit for consistency. I thought putting all those changes in a different PR would be best.
If I misunderstood your question, please elaborate :p

Reply:
Gotcha. Yes, I was wondering if you had plans to test methods like Task() and Update(). I agree, a separate PR sounds like a good idea. Thanks for explaining.